### ISSUE TITLE
Stuck while fitting ensemble

### ISSUE THREAD

**Issue by Motorrat** [automl/auto-sklearn#160]
From time to time I see the below code stuck forever. The root cause is not clear. I have to kill it with CTRL+C. In top I see a python process with zero CPU utilization and very little memory footprint.

```
   seed = 1
    c = AutoSklearnClassifier(
        time_left_for_this_task=300,per_run_time_limit=150,ml_memory_limit=20240,ensemble_size=50,ensemble_nbest=200,
        shared_mode=True, tmp_folder=atsklrn_tempdir, output_folder=atsklrn_tempdir,
        delete_tmp_folder_after_terminate=False, delete_output_folder_after_terminate=False,
        initial_configurations_via_metalearning=0,
        seed=seed)
    c.fit_ensemble(
        task = BINARY_CLASSIFICATION
        ,metric = F1_METRIC
        ,precision = '32'
        ,dataset_name = 'foobar'
        ,ensemble_size=10
        ,ensemble_nbest=15)

    sleep(10)
```

from the logs

```
  File "/x/Redly/anaconda3/lib/python3.5/site-packages/lockfile/linklockfile.py", line 31, in acquire
    os.link(self.unique_name, self.lock_file)
FileExistsError: [Errno 17] File exists: '/x/Redly/truffles/nfs_share/atsklrn_tmp/.auto-sklearn/mgm-19-6811-41172700.20609-2442176699744317723' -> '/x/Redly/truffles/nfs_share/atsklrn_tmp/.auto-sklearn/true_targets_ensemble.npy.lock.lock'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "nfs_share/Redly-SQL-Example2.py", line 449, in <module>
    fit_classifiers(dataframe, pool_size, atsklrn_tempdir, slave_node, predict_only)
  File "nfs_share/Redly-SQL-Example2.py", line 271, in fit_classifiers
    ,ensemble_nbest=15)
  File "/x/Redly/anaconda3/lib/python3.5/site-packages/autosklearn/automl.py", line 480, in fit_ensemble
    self._proc_ensemble.main()
  File "/x/Redly/anaconda3/lib/python3.5/site-packages/autosklearn/ensemble_builder.py", line 94, in main
    targets_ensemble = backend.load_targets_ensemble()
  File "/x/Redly/anaconda3/lib/python3.5/site-packages/autosklearn/util/backend.py", line 117, in load_targets_ensemble
    with lockfile.LockFile(lock_path):
  File "/x/Redly/anaconda3/lib/python3.5/site-packages/lockfile/__init__.py", line 197, in __enter__
    self.acquire()
  File "/x/Redly/anaconda3/lib/python3.5/site-packages/lockfile/linklockfile.py", line 50, in acquire
    time.sleep(timeout is not None and timeout / 10 or 0.1)
KeyboardInterrupt
Traceback (most recent call last):
  File "/x/Redly/anaconda3/lib/python3.5/site-packages/lockfile/linklockfile.py", line 31, in acquire
    os.link(self.unique_name, self.lock_file)
FileExistsError: [Errno 17] File exists: '/x/Redly/truffles/nfs_share/atsklrn_tmp/.auto-sklearn/mgm-19-6811-41172700.20609-2442176699744317723' -> '/x/Redly/truffles/nfs_share/atsklrn_tmp/.auto-sklearn/true_targets_ensemble.npy.lock.lock'

```

**Comment by mfeurer** [comment]
This is indeed an issue, thanks for reporting this. The cause seems to be that a lock file blocks the ensemble process from opening the target values for the validation set. Does this happen when running auto-sklearn in parallel or in sequential mode?

**Comment by Motorrat** [comment]
The fit() method is run in parallel but the ensemble building is sequential - just like the code above. By the time I am getting to the line seed =1 all fit() seeds (should) have already exited. Anyway I do not see their processes. 
Could it be that anaconda3/lib/python3.5/site-packages/lockfile/linklockfile.py is hanging on another lock, not related to autosklearn? (would be my database export code then).

**Comment by mfeurer** [comment]
Okay, I think the issue arises when the target algorithm is killed while it tries to save files to disk. In commit df619b5 I added code which should reduce the likelihood of the target algorithm still locking the file in the later stages of the optimization process. Also, I added increased the time limit for the dummy configuration, which is responsible for saving this file to disk. Please reopen this issue if this problem persists with the upcoming bugfix release.

**Comment by kliesis** [comment]
I have just installed the latest version (0.14.6), unfortunately, get a similar error:

python example_classification.py 

Traceback (most recent call last):
  File "example_classification.py", line 35, in <module>
    automl.fit(X_train, y_train, dataset_name='breast_cancer')
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/estimators.py", line 1051, in fit
    dataset_name=dataset_name,
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/estimators.py", line 375, in fit
    self.automl_.fit(load_models=self.load_models, **kwargs)
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl.py", line 2064, in fit
    is_classification=True,
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl.py", line 614, in fit
    self._backend = self._create_backend()
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl.py", line 287, in _create_backend
    delete_tmp_folder_after_terminate=self._delete_tmp_folder_after_terminate,
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl_common/common/utils/backend.py", line 40, in create
    prefix=prefix,
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl_common/common/utils/backend.py", line 100, in __init__
    self.create_directories()
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl_common/common/utils/backend.py", line 125, in create_directories
    os.makedirs(self.temporary_directory, exist_ok=exist_ok)
  File "/home/zane/.conda/envs/automl/lib/python3.7/os.py", line 223, in makedirs
    mkdir(name, mode)
FileExistsError: [Errno 17] File exists: '/tmp/autosklearn_classification_example_tmp'




If I copy the backend.py from issue #160 to my backend.py, and install lockfile, I get the following error:

Exception ignored in: <function BackendContext.__del__ at 0x7fed821b7a70>
Traceback (most recent call last):
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl_common/common/utils/backend.py", line 77, in __del__
    self.delete_directories(force=False)
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl_common/common/utils/backend.py", line 80, in delete_directories
    if self.delete_output_folder_after_terminate or force:
AttributeError: 'BackendContext' object has no attribute 'delete_output_folder_after_terminate'
Traceback (most recent call last):
  File "example_classification.py", line 35, in <module>
    automl.fit(X_train, y_train, dataset_name='breast_cancer')
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/estimators.py", line 1051, in fit
    dataset_name=dataset_name,
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/estimators.py", line 375, in fit
    self.automl_.fit(load_models=self.load_models, **kwargs)
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl.py", line 2064, in fit
    is_classification=True,
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl.py", line 614, in fit
    self._backend = self._create_backend()
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl.py", line 287, in _create_backend
    delete_tmp_folder_after_terminate=self._delete_tmp_folder_after_terminate,
  File "/home/zane/.conda/envs/automl/lib/python3.7/site-packages/autosklearn/automl_common/common/utils/backend.py", line 27, in create
    prefix=prefix)
TypeError: __init__() got an unexpected keyword argument 'prefix'


How should I best precede? Thanks

**Comment by eddiebergman** [comment]
Hi @Motorrat, this error `FileExistsError: [Errno 17] File exists: '/tmp/autosklearn_classification_example_tmp'` happens because you run the example multiple times and it is set not to delete the folder after use.

Unfortunatly we don't support using the same folder for multiple runs. Our solution is either:
* To give **no** `tmp` folder, which means a random `tmp` folder is used.
* If you want to give an explicit path for autosklearn to work in then you will have to control how it is managed outside of autosklearn's internal code. For example, you may want to check if `os.exists(path)` then `shutil.rmtree(path)` to ensure autosklearn has a clean directory to work from.

In summary, if you're just trying things out for an example or a small experiment, you don't need to provide a path. If you need more control, you should provide a path and manage it.

However for the examples, we should probably not give explicit paths, that's our bad.

Best,
Eddie
